### PR TITLE
[mempool] Improve transaction already in mempool messaging

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
"Transaction already in mempool" is not very helpful of a message, because it doesn't tell you why it failed.  It's also not hiding anything since you can lookup the pending transaction via the API.

So, I've added a breakdown that tells you if the message matches accordingly.

### Test Plan
E2E tests

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1952)

### CODE REVIEW

File: mempool/src/core_mempool/transaction_store.rs [hunk 1]
```diff
@@ -119,29 +119,43 @@ impl TransactionStore {
...
-                    }
+                    };
+                } else if current_version.get_gas_price() > txn.get_gas_price() {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate).with_message(
+                        "Transaction already in mempool with higher gas price".to_string(),
+                    );
                 } else {
-                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate)
-                        .with_message("Transaction already in mempool".to_string());
+                    // If the transaction is the same, it's an idempotent call
```
Comment (JoshLind) [inline mempool/src/core_mempool/transaction_store.rs:156]: nit "call" -> "call."

**Review by JoshLind — APPROVED** [review]
The new error messages are much nicer 😄 Odd about the unused amount though...

File: mempool/src/core_mempool/transaction_store.rs [hunk 2]
```diff
@@ -119,29 +119,43 @@ impl TransactionStore {
...
-                    if let Some(txn) = txns.remove(&txn.sequence_info.transaction_sequence_number) {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate).with_message(
+                        "Transaction already in mempool with different expiration timestamp"
+                            .to_string(),
+                    );
+                } else if current_version.txn.max_gas_amount() != txn.txn.max_gas_amount() {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate).with_message(
+                        "Transaction already in mempool with different max gas amount".to_string(),
+                    );
+                } else if current_version.txn.gas_unit_price() < txn.get_gas_price() {
```
Comment (sitalkedia) [inline mempool/src/core_mempool/transaction_store.rs:146]: These set of `if` and `else-if` statements are a bit hard to follow - may be a cleaner approach would be breaking them down into individual functions? Like `check_for_payload`, `check_for_expiration_time` etc?

**Review by sitalkedia — COMMENTED** [review]
LGTM % minor nits.

File: mempool/src/core_mempool/transaction_store.rs [hunk 3]
```diff
@@ -119,29 +119,43 @@ impl TransactionStore {
...
-                    && current_version.get_gas_price() < txn.get_gas_price()
+                if current_version.txn.payload() != txn.txn.payload() {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate).with_message(
+                        "Transaction already in mempool with different payload".to_string(),
+                    );
+                } else if current_version.txn.expiration_timestamp_secs()
+                    != txn.txn.expiration_timestamp_secs()
                 {
-                    if let Some(txn) = txns.remove(&txn.sequence_info.transaction_sequence_number) {
+                    return MempoolStatus::new(MempoolStatusCode::InvalidUpdate).with_message(
```
[the remark this answers is not in the document]
Comment (gregnazario) [reply mempool/src/core_mempool/transaction_store.rs:138]: ask @zekun000, but I'm assuming that could mean someone could just keep on updating their expiration time and make the mempool a litle more unfair.

File: mempool/src/core_mempool/transaction_store.rs (hunk 2 above)
Comment (bchocho) [inline mempool/src/core_mempool/transaction_store.rs:146]: Should this just be `current_version.get_gas_price() < txn.get_gas_price()`? I see the code is equivalent, so maybe a nit, but it makes me nervous that it's not the same as the next else if.

Comment (gregnazario) [reply mempool/src/core_mempool/transaction_store.rs:146]: This is what I was asking about earlier.  The only difference between the previous code and this is not this if, but now the Authenticator doesn't have to be the same if the txn has the same contents.

The logic before wasn't as explicit and clear about this

File: mempool/src/core_mempool/transaction_store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zekun000) [reply mempool/src/core_mempool/transaction_store.rs:138]: we only allow update to gas afaiu
